### ISSUE TITLE
Too few results from intersect in version v2.26.0-125-g52db654

### ISSUE THREAD

**Issue by mce1** [arq5x/bedtools2#553]
Hi there,

This might pertain to both bedtools and pybedtools so I cc @daler.

I'm trying to select variants in a vcf file that are covered by reads in a bam file. The files I need to process in the end are quite large and I have split them by chromosome so I can use multiprocessing.

Using some test data I have encountered the following error message:

```
mce@:~/mini-example$ bedtools -version
bedtools v2.26.0
mce@laptop:~/mini-example$ python
Python 2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:09:15) 
>>> import pybedtools
>>> variants = pybedtools.BedTool('chromosome_vcf_files/variants_chr1.vcf.gz')
>>> reads = pybedtools.BedTool('chromosome_bam_files/reads_chr1.bam')
>>> covered_variants = variants.intersect(reads, split=True, u=True)
>>> for variant in covered_variants: print variant
... 
chr1	216	rs01	T	G	.	.	RS=1;RPOS=216

chr1	576	rs02	C	A	.	.	RS=2;RPOS=576

chr1	836	rs04	A	C	.	.	RS=4;RPOS=836

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pybedtools/cbedtools.pyx", line 787, in pybedtools.cbedtools.IntervalIterator.__next__ (pybedtools/cbedtools.cxx:11123)
  File "pybedtools/cbedtools.pyx", line 652, in pybedtools.cbedtools.create_interval_from_list (pybedtools/cbedtools.cxx:9208)
IndexError: list index out of range
```

Note that the intersect command itself apparently works fine; only when iterating over the results the error is raised.

Interestingly, with the test data, it appears this error can be bypassed by calling the stream=True option in pybedtools.  

```
>>> covered_variants = variants.intersect(reads, split=True, u=True, stream=True)
>>> for variant in covered_variants: print variant
... 
chr1	216	rs01	T	G	.	.	RS=1;RPOS=216

chr1	576	rs02	C	A	.	.	RS=2;RPOS=576

chr1	836	rs04	A	C	.	.	RS=4;RPOS=836
```

Using the sorted option (as I would like later on with the real data) does not make any difference.

```
>>> covered_variants = variants.intersect(reads, split=True, u=True, sorted=True)
>>> for variant in covered_variants: print variant
... 
chr1	216	rs01	T	G	.	.	RS=1;RPOS=216

chr1	576	rs02	C	A	.	.	RS=2;RPOS=576

chr1	836	rs04	A	C	.	.	RS=4;RPOS=836

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pybedtools/cbedtools.pyx", line 787, in pybedtools.cbedtools.IntervalIterator.__next__ (pybedtools/cbedtools.cxx:11123)
  File "pybedtools/cbedtools.pyx", line 652, in pybedtools.cbedtools.create_interval_from_list (pybedtools/cbedtools.cxx:9208)
IndexError: list index out of range
```

As the error message read similar to the one reported here https://github.com/daler/pybedtools/issues/207 (although I cannot see how groupby is involved in my case) I tried upgrading to the latetst developmental version of bedtools. Indeed, that solved the problem, and, as far as my test example is concerned, everything worked fine now. 

```
mce@laptop:~/mini-example$ bedtools -version
bedtools v2.26.0-125-g52db654
mce@laptop:~/mini-example$ python
Python 2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:09:15)
>>> import pybedtools
>>> variants = pybedtools.BedTool('chromosome_vcf_files/variants_chr1.vcf.gz')
>>> reads = pybedtools.BedTool('chromosome_bam_files/reads_chr1.bam')
>>> covered_variants = variants.intersect(reads, split=True, u=True)
>>> for variant in covered_variants: print variant
... 
chr1	216	rs01	T	G	.	.	RS=1;RPOS=216

chr1	576	rs02	C	A	.	.	RS=2;RPOS=576

chr1	836	rs04	A	C	.	.	RS=4;RPOS=836

>>> covered_variants = variants.intersect(reads, split=True, u=True, sorted=True)
>>> for variant in covered_variants: print variant
... 
chr1	216	rs01	T	G	.	.	RS=1;RPOS=216

chr1	576	rs02	C	A	.	.	RS=2;RPOS=576

chr1	836	rs04	A	C	.	.	RS=4;RPOS=836

```

However, when I used the above developmental version on my real data, intersect gave me far less results than expected.

Using v2.26.0:

```
[mce@znx-cn03 test]$ bedtools -version
bedtools v2.26.0
[mce@znx-cn03 test]$ bedtools intersect -a chromosome_vcf_files/variants_1.vcf.gz -b chromosome_bam_files/reads_1.bam -u -split -sorted | wc -l
723917
```

Upon upgrading to v2.26.0-125-g52db654:

```
[mce@znx-cn03 test]$ bedtools -version
bedtools v2.26.0-125-g52db654
[mce@znx-cn03 test]$ bedtools intersect -a chromosome_vcf_files/variants_1.vcf.gz -b chromosome_bam_files/reads_1.bam -u -split -sorted | wc -l
0
```

For some of the chromosome subfiles some variants are reported with v2.26.0-125-g52db654 but the number is always far lower than what I expect. When I look at files with multiple chromosomes, typically all the hits come from one or a few chromosomes with no overlapping reads/variants reported for the others. 

As a work around I tried to use v2.26.0 with the stream=True option I mentioned above, however, for these very large files this does not work but I get the same error as before:

```
[mce@znx-cn03 test]$ python test.py > tmp.txt
 Traceback (most recent call last):
  File "basic.py", line 880, in <module>
    heterozygous_snps_by_chromosome = pool.map(get_heterozygous_snps_as_bedfile, reads_SQ_lines)
  File "/home/mengeholm/anaconda2/lib/python2.7/multiprocessing/pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/home/mengeholm/anaconda2/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
IndexError: list index out of range
```

So, in summary, I suspect there is a problem with intersect in version v2.26.0-125-g52db654 (which is not present in v2.26.0). I am not sure what the IndexError means when iterating over the results of intersect in v2.26.0, specifically, if that is due to a problem with bedtools or pybedtools and if it can be fixed on the pybedtools side maybe.

I would be really grateful if (one of) you could have a look.

Many thanks in advance,

mce

**Comment by daler** [comment]
@mce1 the index error is certainly from pybedtools but the error could be caused from a malformed line from input or intersect output, or some other issue with pybedtools.  Would you please provide example files that reproduce these errors?

**Comment by mce1** [comment]
@daler please find the two files attached

[variants_chr1.vcf.gz](https://github.com/arq5x/bedtools2/files/1191844/variants_chr1.vcf.gz)

[reads_chr1.sam.txt](https://github.com/arq5x/bedtools2/files/1191857/reads_chr1.sam.txt)

(The bam file I have converted to .sam.txt because otherwise it would not upload.)
Thanks, mce

**Comment by mce1** [comment]
@daler can you reproduce the IndexError with the example files above?

**Comment by daler** [comment]
@mce1 sorry was on vacation and just got back. Here is my setup, I'm unable to reproduce the IndexError. I do see 2.26 outputting zero lines, though it appears you saw dev outputting zero lines. Do you see it using the following setup? 

Contents of `ex.py`:

```python
import pybedtools
import subprocess as sp
print('pbt:', pybedtools.__version__)
print(
    sp.check_output(['which', 'bedtools'], universal_newlines=True).strip(),
    '(',
    sp.check_output(['bedtools', '--version'], universal_newlines=True).strip(),
    ') says:'
)
a = pybedtools.BedTool('variants_chr1.vcf.gz')
b = pybedtools.BedTool('reads_chr1.bam')
c = a.intersect(b, u=True, split=True, sorted=True)
for i in c:
    print(i, end='')
print()
```

Contents of `test.sh`:

```bash
#!/bin/bash

HERE=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )

if [[ ! -e bedtools-v2.26/bin/bedtools ]]; then
    git clone https://github.com/arq5x/bedtools2.git bedtools-v2.26
    cd bedtools-v2.26
    git checkout v2.26.0
    make
fi


if [[ ! -e bedtools-dev/bin/bedtools ]]; then
    git clone https://github.com/arq5x/bedtools2.git bedtools-dev
    cd bedtools-dev
    make
fi

if [[ ! -e reads_chr1.bam ]]; then
    wget https://github.com/arq5x/bedtools2/files/1191844/variants_chr1.vcf.gz
    wget -O- https://github.com/arq5x/bedtools2/files/1191857/reads_chr1.sam.txt > reads_chr1.sam
    samtools view -Sb reads_chr1.sam > reads_chr1.bam
fi

PATH="$HERE/bedtools-dev/bin:$PATH" python ex.py
PATH="$HERE/bedtools-v2.26/bin:$PATH" python ex.py
```

Output:

```
pbt: 0.7.10
/tmp/bedtools-dev/bin/bedtools ( bedtools v2.26.0-125-g52db6549 ) says:
chr1    216     rs01    T       G       .       .       RS=1;RPOS=216
chr1    576     rs02    C       A       .       .       RS=2;RPOS=576
chr1    836     rs04    A       C       .       .       RS=4;RPOS=836

pbt: 0.7.10
/tmp/bedtools-v2.26/bin/bedtools ( bedtools v2.26.0 ) says:

```

**Comment by mce1** [comment]
@daler: thanks a lot for the reply! 

I can reproduce this: using bedtools 2.26 I now see an empty list instead of the IndexError. I think this is most likely due to the different versions of python, i.e. python3 now vs. python2 before. Can you reproduce the IndexError with python 2.7?

The bedtools dev version gives the same results as before, but as I said before there seems to be another problem with this version which does not become apparent with the test files here but only with larger data sets where intersect produces far fewer results than in version 2.26. This has nothing to do with pybedtools of course.

**Comment by daler** [comment]
@mce1 I've updated the testing scripts to check py2 and py3, and also run the same commands outside the context of pybedtools, see this gist:

https://gist.github.com/daler/9623fad6b3cfa1c599591b0538e9f4dd

I was able to reproduce the IndexError in py2 you originally reported, thanks for pointing that out. It appears to be a difference in behavior between py2 and py3 on malformed files.

As you can see from the files created, the v2.26 version has binary information at the end. Looks like it's related to https://github.com/arq5x/bedtools2/issues/497. As far as line counts, it is possible that when you checked the line count on your real data on v2.26 results, `wc -l` was picking up newline characters in the binary section, and so reported an inaccurate value. You could try `grep "^chr" output.vcf | wc -l` (assuming all chroms start with "chr") for a potentially more accurate count in that bedtools version.

I'm afraid there's not much I can do on the pybedtools side, I need to pass it off to @arq5x.

**Comment by mce1** [comment]
@daler and @arq5x,

thanks a lot for your help with this so far. Below, I have attached another example which yields unexpected results with intersect in the dev version, this time when trying to get variants from a vcf that overlap transcripts in a bed file. I'm not entirely sure that this reflects the same problem I had originally (when trying to extract variants that overlap reads in bam format) but maybe it can be useful for debugging.

contents of `test.py`:

```
from __future__ import print_function
import pybedtools

import subprocess as sp
print('pbt:', pybedtools.__version__)
pybedtools.debug_mode(True)
print(
    sp.check_output(['which', 'bedtools'], universal_newlines=True).strip(),
    '(',
    sp.check_output(['bedtools', '--version'], universal_newlines=True).strip(),
    ') says:'
)

def gene_transcript_names_dictionary_from_table(filename):
    table = open(filename, 'r')
    dictionary = {}
    for line in table:
        if line[0] != "#":
            if line.split('\t')[5].rstrip() + '_' + line.split('\t')[1].rstrip() in dictionary:
                dictionary[line.split('\t')[5].rstrip() + '_' + line.split('\t')[1].rstrip()].append(line.split('\t')[0])
            else:
                dictionary[line.split('\t')[5].rstrip() + '_' + line.split('\t')[1].rstrip()] = [line.split('\t')[0]]
    return dictionary

def rename_chromosomes(feature):
    if "chr" in feature.chrom: feature.chrom = feature.chrom.replace("chr", "")
    else: feature.chrom = "chr" + feature.chrom
    return feature

def get_BedTool_transcripts_by_gene(gene):
    gene_transcript_names = gene_transcript_names_dictionary[gene]
    return [Interval for Interval in BedTool_transcripts_bed_all if Interval.name in gene_transcript_names]

def get_BedTool_variants_by_gene(gene):
    chromosome = gene.split('_chr')[1]
    chromosome_variants = pybedtools.BedTool('chr%s.vcf.gz' % chromosome)
    gene_variants_vcf = chromosome_variants.intersect(gene_BedTool_transcripts_bed_dictionary[gene], split=True, u=True) 
    print(gene + ':')
    print(len(gene_variants_vcf))
    for variant in gene_variants_vcf:
        print(variant)
    return

gene_transcript_names_dictionary = gene_transcript_names_dictionary_from_table('app_snca.gene_table.txt')

all_genes = gene_transcript_names_dictionary.keys()

BedTool_transcripts_bed_all = pybedtools.BedTool('app_snca.bed.gz').each(rename_chromosomes).saveas()

gene_BedTool_transcripts_bed_dictionary = dict(zip(all_genes,
                                                   map(get_BedTool_transcripts_by_gene, all_genes)))

gene_BedTool_variants_dictionary = dict(zip(all_genes,
                                                   map(get_BedTool_variants_by_gene, all_genes)))

```

contents of `test.sh`:

```
#!/bin/bash                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                         
HERE=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                         
# build python 2 and python 3 environments                                                                                                                                                                                                                               
if [[ ! -e py2env ]]; then                                                                                                                                                                                                                                               
    conda create -y --prefix py2env pybedtools python=2 -c conda-forge -c bioconda                                                                                                                                                                                       
fi                                                                                                                                                                                                                                                                       
if [[ ! -e py3env ]]; then                                                                                                                                                                                                                                               
    conda create -y --prefix py3env pybedtools python=3 -c conda-forge -c bioconda                                                                                                                                                                                       
fi                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                         
if [[ ! -e bedtools-v2.26/bin/bedtools ]]; then                                                                                                                                                                                                                          
    git clone https://github.com/arq5x/bedtools2.git bedtools-v2.26                                                                                                                                                                                                      
    (cd bedtools-v2.26 && git checkout v2.26.0 && make)                                                                                                                                                                                                                  
fi                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                         
if [[ ! -e bedtools-dev/bin/bedtools ]]; then                                                                                                                                                                                                                            
    git clone https://github.com/arq5x/bedtools2.git bedtools-dev                                                                                                                                                                                                        
    (cd bedtools-dev && make)                                                                                                                                                                                                                                            
fi                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                         
for py in 2 3; do                                                                                                                                                                                                                                                        
    source activate py${py}env                                                                                                                                                                                                                                           
    for bt in dev v2.26; do                                                                                                                                                                                                                                              
        echo                                                                                                                                                                                                                                                             
        echo "==========================="                                                                                                                                                                                                                               
        echo "Python = py$py"                                                                                                                                                                                                                                            
        echo "BEDTools = $bt"                                                                                                                                                                                                                                            
        PATH="$HERE/bedtools-$bt/bin:$PATH" python test.py                                                                                                                                                                                                                 
    done                                                                                                                                                                                                                                                                 
    source deactivate         
done

```

Test files:
[app_snca.gene_table.txt](https://github.com/arq5x/bedtools2/files/1254351/app_snca.gene_table.txt)
[app_snca.bed.gz](https://github.com/arq5x/bedtools2/files/1254352/app_snca.bed.gz)
[chr6.vcf.gz](https://github.com/arq5x/bedtools2/files/1254353/chr6.vcf.gz)
[chr16.vcf.gz](https://github.com/arq5x/bedtools2/files/1254354/chr16.vcf.gz)


From what I can see, python2 and python3 give the same results here. In each case, v26.0 yields 17 exonic variants for one of the two genes (app on chromosome 16) and 15 for the other (snca on chromosome 6) as expected. The dev version, on the other hand, returns all the variants in the test vcf (2706) for the first gene but not a single one for the second gene.

Would be great if you could have a look at this as well as the original problem described above.

Many thanks in advance,

mce

**Comment by arq5x** [comment]
I will try to dig into this today!

**Comment by arq5x** [comment]
Thanks for reporting this - the bug will require some substantial refactoring and testing and may take awhile to resolve. My apologies.

**Comment by arq5x** [comment]
Thanks to help from @nkindlon this is now fixed in master.
